source/extensions/filters/network/postgres_proxy/postgres_decoder.cc

```diff
@@ -176,134 +179,216 @@ void DecoderImpl::initialize() {
+    return Decoder::Result::NeedMoreData;
+  }
+
+  // Validate the message before processing.
+  const MsgBodyReader& f = std::get<1>(first_);
+  const auto msgParser = f();
+  // Run the validation.
+  message_len_ = data.peekBEInt<uint32_t>(0);
+// MAX_STARTUP_PACKET_LENGTH is defined in Postgres source code
+// as maximum size of initial packet.
+#define MAX_STARTUP_PACKET_LENGTH 10000
+  if (message_len_ > MAX_STARTUP_PACKET_LENGTH) {
+    // Message does not conform to the expected format. Move to out-of-sync state.
+    data.drain(data.length());
+    state_ = State::OutOfSyncState;
+    return Decoder::Result::ReadyForNext;
   }
 
-  if (!startup_) {
-    data.copyOut(0, 1, &command_);
-    ENVOY_LOG(trace, "postgres_proxy: command is {}", command_);
+  Message::ValidationResult validationResult = msgParser->validate(data, 4, message_len_ - 4);
+
+  if (validationResult == Message::ValidationNeedMoreData) {
+    return Decoder::Result::NeedMoreData;
   }
 
-  // The 1 byte message type and message length should be in the buffer
-  // Check if the entire message has been read.
-  std::string message;
-  message_len_ = data.peekBEInt<uint32_t>(startup_ ? 0 : 1);
-  if (data.length() < (message_len_ + (startup_ ? 0 : 1))) {
-    ENVOY_LOG(trace, "postgres_proxy: cannot parse message. Need {} bytes in buffer",
-              message_len_ + (startup_ ? 0 : 1));
-    // Not enough data in the buffer.
-    return Decoder::NeedMoreData;
+  if (validationResult == Message::ValidationFailed) {
+    // Message does not conform to the expected format. Move to out-of-sync state.
+    data.drain(data.length());
+    state_ = State::OutOfSyncState;
+    return Decoder::Result::ReadyForNext;
   }
 
-  if (startup_) {
-    uint32_t code = data.peekBEInt<uint32_t>(4);
-    // Startup message with 1234 in the most significant 16 bits
-    // indicate request to encrypt.
-    if (code >= 0x04d20000) {
-      encrypted_ = true;
-      // Handler for SSLRequest (Int32(80877103) = 0x04d2162f)
-      // See details in https://www.postgresql.org/docs/current/protocol-message-formats.html.
-      if (code == 0x04d2162f) {
-        // Notify the filter that `SSLRequest` message was decoded.
-        // If the filter returns true, it means to pass the message upstream
-        // to the server. If it returns false it means, that filter will try
-        // to terminate SSL session and SSLRequest should not be passed to the
-        // server.
-        encrypted_ = callbacks_->onSSLRequest();
-      }
-
-      // Count it as recognized frontend message.
-      callbacks_->incMessagesFrontend();
-      if (encrypted_) {
-        ENVOY_LOG(trace, "postgres_proxy: detected encrypted traffic.");
-        incSessionsEncrypted();
-        startup_ = false;
-      }
-      data.drain(data.length());
-      return encrypted_ ? Decoder::ReadyForNext : Decoder::Stopped;
+  Decoder::Result result = Decoder::Result::ReadyForNext;
+  uint32_t code = data.peekBEInt<uint32_t>(4);
```

Comment:
Rename `code` to `startup_packet` just to use the Postgres vocabulary.

Reply:
Yeah, but this is not really a packet itself, but just 4-bytes code which can be either a version or encryption type.